### PR TITLE
[tests-only][full-ci] Add test coverage for file preview using space dav version

### PR DESCRIPTION
**Pull request by amrita-shrestha**
## Description
Previously there was test coverage for previewing resources using new wed-dav only. This PR has added tests coverage of preview using old and spaces wed-dav version in personal spaces.
- https://github.com/owncloud/ocis/issues/2071
 decision let error different in ocis and oc10 : https://github.com/owncloud/ocis/issues/2071#issuecomment-1655370760
- https://github.com/owncloud/ocis/issues/2064 (Fixed in master branch so skipped on stable branch)

Open task:
- [ ] add tests coverage for previewing shared resource in spaces webdav

## Related Issue
- Part of https://github.com/owncloud/ocis/issues/6752

## How Has This Been Tested?
- Locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/TestHelpers/WebDavHelper.php
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:883]:
```suggestion
```

File: tests/TestHelpers/WebDavHelper.php
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:901]: can they be null?
```suggestion
		string $property,
		?string $baseUrl,
		?string $user,
		?string $password,
		?string $path,
```

File: tests/TestHelpers/WebDavHelper.php
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:929]:
```suggestion
                Assert::strictEquals(207, $statusCode, "HTTP status code should be 200 but found $statusCode\nResponse Body:\n$responseBody");
```

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:469]: `$url` can be used like this
```suggestion
                $url = $this->getBaseUrl() . '/remote.php';
		$davVersion = $this->getDavPathVersion();
		if ($davVersion === WebDavHelper::DAV_VERSION_SPACES) {
			$fileId = urlencode(strstr($this->getFileIdForPath($doDavRequestAsUser ?? $user, $path), '!', true));

			$eTag = str_replace('"', '', $this->getFileEtagForPath($doDavRequestAsUser ?? $user, $path));
			$url .= '/dav/spaces/' . $fileId . '/';
```

File: tests/TestHelpers/WebDavHelper.php
Comment (SagarGi) [inline tests/TestHelpers/WebDavHelper.php:895]: what was wrong in the previous function and creating new function since the implementation looks same to me?

Comment (amrita-shrestha) [reply tests/TestHelpers/WebDavHelper.php:895]: its more dynamic then before can use for all type of property rather than fileid

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/WebDav.php:4544]: instead of creating new step cannot we reuse the old one and adjust the `dav-verison` code there?

File: tests/TestHelpers/WebDavHelper.php
Comment (SagarGi) [reply tests/TestHelpers/WebDavHelper.php:895]: if so then may be only change the function name no need to create new function.

File: tests/TestHelpers/WebDavHelper.php
Comment (grgprarup) [reply tests/TestHelpers/WebDavHelper.php:929]:
```suggestion
		if ($statusCode != 207) {
			throw new Error($responseBody . "\n" . "HTTP status code should be 200 but found $statusCode");
		}
```
The status code is compared with `207` but in the message `status code should be 200`. why?

File: tests/acceptance/expected-failures-API-on-OCIS-storage.md
Comment (grgprarup) [inline tests/acceptance/expected-failures-API-on-OCIS-storage.md:219]: IMHO, Let's not add this to expected failure. The failure is with a response message which can be easily fixed. Let's get the fix done and refactor the message in the new PR.

File: tests/TestHelpers/WebDavHelper.php
Comment (amrita-shrestha) [reply tests/TestHelpers/WebDavHelper.php:895]: not in this PR

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:4544]: this preview one is a little bit complex to adjust in old code. As you can see  we dont need space id here like other space dav version

File: tests/acceptance/expected-failures-API-on-OCIS-storage.md
Comment (saw-jan) [reply tests/acceptance/expected-failures-API-on-OCIS-storage.md:219]: Yeah. Let dev team push a fix PR for https://github.com/owncloud/ocis/issues/2071

CC @ScharfViktor @micbar @2403905 
or any opinions?

Comment (ScharfViktor) [reply tests/acceptance/expected-failures-API-on-OCIS-storage.md:219]: #2071 has stayed open for over 2 years. I will inform dev team today.

Comment (saw-jan) [reply tests/acceptance/expected-failures-API-on-OCIS-storage.md:219]: This could be the chance to cleanup that issue :smile:

**Comment by SagarGi** [comment]
This PR is blocked by this issue: https://github.com/owncloud/ocis/issues/2071

File: tests/acceptance/features/coreApiWebdavPreviews/previews.feature
Comment (saw-jan) [inline tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: how to tell that we are trying to access file of Alice while using old dav? I think we can remote old from here
```suggestion
```

Comment (saw-jan) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: Also, I doubt if preview works with old webdav.

Comment (saw-jan) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: It's just because we don't have test coverage for old webdav in core.

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:478]:
```suggestion
			$urlParameter['a'] = '1';
			$urlParameter['c'] = $eTag;
```

File: tests/acceptance/features/coreApiWebdavPreviews/previews.feature
Comment (saw-jan) [inline tests/acceptance/features/coreApiWebdavPreviews/previews.feature:161]: dev team has decided to keep the oCIS error message
```suggestion
    And the value of the item "/d:error/s:message" in the response about user "Alice" should be "File with name parent.txt could not be located"
```

**Review by saw-jan — CHANGES_REQUESTED** [review]
See https://github.com/owncloud/ocis/issues/2071#issuecomment-1655370760

File: tests/acceptance/features/coreApiWebdavPreviews/previews.feature
Comment (amrita-shrestha) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: @saw-jan 
can you first clear confusion and comment here because it creates a lot of confusion for others

Comment (saw-jan) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: It is marked resolved. So why confusion and I have made myself clear there.

Comment (saw-jan) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: Old webdav doesn't make sense in this particular scenario so the the review is here.

Comment (saw-jan) [reply tests/acceptance/features/coreApiWebdavPreviews/previews.feature:165]: About the download preview via old webadv, I have edited myself and made that clear in other comment. so that others also don't get confused.

File: tests/acceptance/expected-failures-API-on-OCIS-storage.md
Comment (saw-jan) [inline tests/acceptance/expected-failures-API-on-OCIS-storage.md:222]: totally forgot that the fix was in master. but IMO, we have to put tests here for stable

File: tests/acceptance/expected-failures-API-on-OCIS-storage.md
Comment (saw-jan) [inline tests/acceptance/expected-failures-API-on-OCIS-storage.md:566]: Fixed in https://github.com/owncloud/ocis/pull/6932

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:480]: do we really need etag?

File: tests/TestHelpers/WebDavHelper.php
Comment (SwikritiT) [inline tests/TestHelpers/WebDavHelper.php:931]:
```suggestion
			throw new Error($responseBody . "\n" . "HTTP status code should be 207 but found $statusCode");
```

File: tests/TestHelpers/WebDavHelper.php
Comment (SwikritiT) [reply tests/TestHelpers/WebDavHelper.php:87]: for when we need to get the body for different properties dynamically there `getBodyForPropfind()` which can be used IMO

Comment (amrita-shrestha) [reply tests/TestHelpers/WebDavHelper.php:87]: @SwikritiT I forget to mention in the description of this PR I am trying to avoid maximum changes. In follow-up PR I will implement DRY concept related to getFilePropertyForPath. `getFilePropertyForPath` is introduced so that `getFileIdForPath` method remains untouched and we can reuse the same code for etag too.

In my follow-up PR `getFileIdForPath` and `getFileEtagForPath` will be replaced by `getFilePropertyForPath` . Main concept of `getFilePropertyForPath` is to able to use dynamically 
Maybe I can use propfind in place of this 
```
$body
			= "<?xml version='1.0'?>
				<d:propfind  xmlns:d='DAV:' xmlns:oc='http://owncloud.org/ns'>
 					<d:prop>
 					    <$property />
  					</d:prop>
				</d:propfind>";

		$response = self::makeDavRequest(
			$baseUrl,
			$user,
			$password,
			"PROPFIND",
			$path,
			null,
			$xRequestId,
			$body,
			$davPathVersionToUse
		);
```

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:480]: from my finding and inspecting the browser, we need `etag` but if there is an alternative that I am unaware of. Please share with me.

Comment (individual-it) [reply tests/acceptance/features/bootstrap/WebDav.php:480]: if it does not work without the etag, we need to find out if that is a bug or feature

File: tests/TestHelpers/WebDavHelper.php
Comment (saw-jan) [reply tests/TestHelpers/WebDavHelper.php:87]: etag should not be required to request preview. so we won't be needing this for now

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:480]: viktor is looking over it

Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:480]: @saw-jan We are going to use `etag` confirmed by owncloud side reason mention in the description

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/WebDav.php:480]: so, use it for others too (old and new)

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:473]: Let's not use etag, it will only be extra work for the tests. The server does nothing with `c` parameter as of now.

CC @ScharfViktor
```suggestion
```

Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/WebDav.php:473]: Yes, @amrita-shrestha sorry for the confusion. I double-checked with the backend developers, the backend doesn't use the "c" parameter, so we don't need to include it in apiTest

Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:473]: Do I have to wait until the backend developer removes 'c' parameter or I can move on with this PR @ScharfViktor

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/WebDav.php:473]: No, you don't have to. The backend hasn't made use of `c` (etag) yet so no need to remove anything in the backend. You can continue on this PR.

**Comment by phil-davis** [comment]
This PR is to `stable-3.0` branch. Is this a backport from master branch?

**Comment by SwikritiT** [comment]
> This PR is to `stable-3.0` branch. Is this a backport from master branch?

I think it'll be forward ported after being merged in stable

File: tests/acceptance/features/coreApiWebdavPreviews/previews.feature
Comment (saw-jan) [inline tests/acceptance/features/coreApiWebdavPreviews/previews.feature:151]: for spaces? Also, do we need comments?

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:473]: should we use resource-id or resource-name? I would vote for `resource-name`
`/dav/spaces/<personal-space-id>/$path`

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:4668]: use filename instead of fileid

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (amrita-shrestha) [reply tests/acceptance/features/bootstrap/WebDav.php:473]: cc @ScharfViktor

File: tests/acceptance/features/bootstrap/WebDav.php
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDav.php:4683]: same as `downloadPreviews()`?

**Comment by amrita-shrestha** [comment]
I will make a new PR for stable4.0.0

**Pull request closed without merging**
